### PR TITLE
Feature:: Preprocessing models

### PR DESCRIPTION
**Pull request by AHReccese**
#### Reference Issues/PRs
#88 

#### What does this implement/fix? Explain your changes.
In this PR following changes are applied:
##### Following preprocessing classes are implemented and tested. In front of each module, I wrote the problem we faced when adding it to Pymilo.

+ [x] FunctionTransformer
+ [x] KBinsDiscretizer ( :no_entry:  Shape miss-match `array([[..],...,[...]])` and `array([array([..]),...,array([...])])`
+ [x] KernelCenterer
+ [x] MultiLabelBinarizer
+ [x] MaxAbsScaler
+ [x] Normalizer
+ [x] OrdinalEncoder ( :warning:  when saving the JSON file one of `encoded_missing_value` field could be `NaN` which I think can make JSON file unparsable)
+ [x] PolynomialFeatures
+ [x] PowerTransformer ( :no_entry: Dependent on the `StandardScaler` transportation - transportation loop)
+ [x] QuantileTransformer
+ [x] RobustScaler
+ [x] SplineTransformer ( :no_entry: Dependent on the `scipy.interpolate.BSpline`)
+ [x] TargetEncoder ( :no_entry: `sklearn.__version__ > 1.3`)

##### Tests for the following preprocessing classes are tested:
+ [x] StandardScaler
+ [x] OneHotEncoder
+ [x] LabelBinarizer
+ [x] LabelEncoder

#### Any other comments?

### CODE REVIEW

**Comment by AHReccese** [comment]
@sadrasabouri 
Dear Sadra, please review my updates.

**Comment by AHReccese** [comment]
NaN-type support is also added.
Also please listen to voices in telegram, I've explained what I've done to handle these cases:
- KBinsDiscretizer
- PowerTransformer
- SplineTransformer
- TargetEncoder
- OrdinalEncoder

**Comment by AHReccese** [comment]
We will address the coverage issue in a separate PR.
I will enhance code coverage to over 90%.

File: pymilo/transporters/preprocessing_transporter.py [hunk 1]
```diff
@@ -96,7 +113,71 @@ def deserialize_pre_module(self, serialized_pre_module):
         data = serialized_pre_module["pymilo-preprocessing-data"]
         associated_type = SKLEARN_PREPROCESSING_TABLE[serialized_pre_module["pymilo-preprocessing-type"]]
         retrieved_pre_module = associated_type()
-        gdst = GeneralDataStructureTransporter()
+        for key, _ in data.items():
```
Comment (sadrasabouri) [inline pymilo/transporters/preprocessing_transporter.py:116]: We can also use the `value` in following `if` statement conditions. If you are not using `value`, you can simply iterate over `data`.

**Review by sadrasabouri — CHANGES_REQUESTED** [review]
Thanks Amirhossein. I left minor comments.

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,13 +6,39 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- preprocessing params fulfilled in `pymilo_param`
+- `SUPPORTED_MODELS.md` updated
+- `README.md` updated
+- `serialize_possible_ml_model` in the Ensemble chain
 ## [0.8] - 2024-05-06
 ### Added
 - `StandardScaler` Transformer in `pymilo_param.py`
 - `PreprocessingTransporter` Transporter
 - ndarray shape config in `GeneralDataStructure` Transporter
-- `util.py` in chains
```
[the remark this answers is not in the document]
Comment (AHReccese) [reply CHANGELOG.md:15]: What's the former format of this line?

File: SUPPORTED_MODELS.md [hunk 3]
```diff
@@ -635,4 +635,54 @@
...
+		<td>13</td>
+		<td><b>RobustScaler</b></td>
+		<td>>=0.9</td>
+	</tr>
+	<tr align="center">
+		<td>14</td>
+		<td><b>QuantileTransformer</b></td>
+		<td>>=0.9</td>
+	</tr>
 </table>
```
[the remark this answers is not in the document]
Comment (AHReccese) [reply SUPPORTED_MODELS.md:688]: Sure Sure

File: pymilo/utils/util.py [hunk 4]
```diff
@@ -119,3 +119,21 @@ def has_named_parameter(func, param_name):
...
+    :param a: outer list
+    :type a: list
+    :param b: inner list
+    :type b: list
+
+    :return: boolean
+    """
+    if len(a) < len(b):
+        return False
+    if a[0] == b[0] and all(a[j] == b[j] for j in range(len(b))):
```
[the remark this answers is not in the document]
Comment (AHReccese) [reply pymilo/utils/util.py:137]: you are right.

File: pymilo/transporters/general_data_structure_transporter.py [hunk 5]
```diff
@@ -450,4 +458,17 @@ def deep_deserialize_ndarray(self, deserialized_ndarray):
...
-        return np.asarray(new_list, dtype=dtype).reshape(shape)
+
+        if dtype == "object":
+            item = new_list[0]
+            # check if inner items have specific dtype.
+            try:
+                _ = item.dtype
+                pre_result = np.asarray(new_list)
+            except AttributeError:
+                pre_result = np.asarray(new_list, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (AHReccese) [reply pymilo/transporters/general_data_structure_transporter.py:469]: I will update this part, you're right.

File: pymilo/transporters/preprocessing_transporter.py [hunk 6]
```diff
@@ -4,7 +4,13 @@
 from ..utils.util import check_str_in_iterable, get_sklearn_type
 from .transporter import AbstractTransporter, Command
 from .general_data_structure_transporter import GeneralDataStructureTransporter
+from .function_transporter import FunctionTransporter
+from scipy.interpolate._bsplines import BSpline
```
[the remark this answers is not in the document]
Comment (AHReccese) [reply pymilo/transporters/preprocessing_transporter.py:8]: `scipy` is getting import by scikit implicitly

File: pymilo/transporters/preprocessing_transporter.py (hunk 1 above)
Comment (AHReccese) [reply pymilo/transporters/preprocessing_transporter.py:116]: good point, sure.

**Comment by AHReccese** [comment]
Dear Sadra, I've applied your suggestion, thank you so much for your detailed review. I really appreciate it.
Please re-review again.

File: CHANGELOG.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadrasabouri) [reply CHANGELOG.md:15]:
```
`util.py` in chains
```

[the remark this answers is not in the document]
Comment (AHReccese) [reply CHANGELOG.md:15]: what's the difference? 
@sadrasabouri

[the remark this answers is not in the document]
Comment (sadrasabouri) [reply CHANGELOG.md:15]: You changed it to :
```
`util.py` in `chains`
```

[the remark this answers is not in the document]
Comment (AHReccese) [reply CHANGELOG.md:15]: oooh right, could you please apply these changes by yourself? you are way more accurate than I am :)
I completely agree.
@sadrasabouri

[the remark this answers is not in the document]
Comment (sadrasabouri) [reply CHANGELOG.md:15]: Sure

**Review by sadrasabouri — APPROVED** [review]
LGTM now,
@sepandhaghighi can you please review it?

**Comment by sepandhaghighi** [comment]
> LGTM now, @sepandhaghighi can you please review it?

I will review this pull request within the next 48 hours. 💯

File: pymilo/transporters/preprocessing_transporter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sepandhaghighi) [reply pymilo/transporters/preprocessing_transporter.py:8]: We need to add `scipy` to both `requirements.txt` and `dev-requirements.txt` because there is no guarantee that `scipy` will remain an essential requirement for `scikit-learn` in the future.

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,7 +6,33 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- `RobustScaler` preprocessing model
+- `PolynomialFeatures` preprocessing model
+- `OrdinalEncoder` preprocessing model
+- `Normalizer` preprocessing model
+- `MaxAbsScaler` preprocessing model
+- `MultiLabelBinarizer` preprocessing model
+- `KernelCenterer` preprocessing model
+- `FunctionTransformer` preprocessing model
+- `Binarizer` preprocessing model
+- `tests/test_preprocessings/util.py`
```
Comment (sepandhaghighi) [inline CHANGELOG.md:24]: remove this line

File: SUPPORTED_MODELS.md
Comment (sepandhaghighi) [inline SUPPORTED_MODELS.md:1]: `Last Update: 2024-05-30`

File: pymilo/utils/util.py [hunk 8]
```diff
@@ -119,3 +119,19 @@ def has_named_parameter(func, param_name):
     _signature = signature(func)
     parameter_names = [p.name for p in _signature.parameters.values()]
     return param_name in parameter_names
+
+
+def prefix_list(a, b):
+    """
+    Check whether the b list is a sublist of the a list.
+
+    :param a: outer list
```
Comment (sepandhaghighi) [inline pymilo/utils/util.py:128]: I recommend using `list1` and `list2` instead of `a` and `b`.

**Review by sepandhaghighi — CHANGES_REQUESTED** [review]
@AHReccese, @sadrasabouri Thanks for your efforts! 💯 
Please review my comments. 🔥

File: pymilo/utils/util.py (hunk 8 above)
Comment (sadrasabouri) [reply pymilo/utils/util.py]: fixed in [a600f62](https://github.com/openscilab/pymilo/pull/104/commits/a600f6272c3a50422d6ef57f75d124ba6c63694a).

File: CHANGELOG.md (hunk 7 above)
Comment (sadrasabouri) [reply CHANGELOG.md]: fixed in [3d6eacb](https://github.com/openscilab/pymilo/pull/104/commits/3d6eacb5a617611e808f9fb2aa07f9e3c8d8c701).

File: pymilo/transporters/preprocessing_transporter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sadrasabouri) [reply pymilo/transporters/preprocessing_transporter.py:8]: fixed [4758dcf](https://github.com/openscilab/pymilo/pull/104/commits/4758dcfe873716c9f34417714750a5905a7aea73).

**Comment by AHReccese** [comment]
@sadrasabouri 
Thank you Sadra for applying Sepand's comments, I went through them one by one and it is great!
@sepandhaghighi waiting for your final review sir.

**Review by sepandhaghighi — APPROVED** [review]
LGTM 🔥
